source/common/filter/ext_authz.cc

```diff
@@ -0,0 +1,101 @@
+#include "common/filter/ext_authz.h"
+
+#include <cstdint>
+#include <string>
+
+#include "common/common/assert.h"
+#include "common/tracing/http_tracer_impl.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+namespace TcpFilter {
+
+InstanceStats Config::generateStats(const std::string& name, Stats::Scope& scope) {
+  const std::string final_prefix = fmt::format("ext_authz.{}.", name);
+  return {ALL_TCP_EXT_AUTHZ_STATS(POOL_COUNTER_PREFIX(scope, final_prefix),
+                                  POOL_GAUGE_PREFIX(scope, final_prefix))};
+}
+
+void Instance::callCheck() {
+  CheckRequestUtils::createTcpCheck(filter_callbacks_, checkRequest_);
+
+  status_ = Status::Calling;
+  filter_callbacks_->connection().readDisable(true);
+  config_->stats().active_.inc();
+  config_->stats().total_.inc();
+
+  calling_check_ = true;
+  client_->check(*this, checkRequest_, Tracing::NullSpan::instance());
+  calling_check_ = false;
+}
+
+Network::FilterStatus Instance::onData(Buffer::Instance&) {
+  if (status_ == Status::NotStarted) {
+    // If there was no ssl handshake there will be no event Network::ConnectionEvent::Connected
+    // and onData() will be the first event into the filter.
+    // Therefore, if the authorization service hasn not been invoked then we must do so now.
+    callCheck();
+  }
+  return status_ == Status::Calling ? Network::FilterStatus::StopIteration
+                                    : Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus Instance::onNewConnection() {
+  // Wait till the next event occurs.
+  return Network::FilterStatus::Continue;
+}
+
+void Instance::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    if (status_ == Status::Calling) {
+      // Make sure that any pending request in the client is cancelled. This will be NOP if the
+      // request already completed.
+      client_->cancel();
+      config_->stats().active_.dec();
+    }
+  } else {
+    // SSL connection is post TCP newConnection. Therefore the ext_authz check in onEvent.
+    // if the ssl handshake was successful then it will invoke the
+    // Network::ConnectionEvent::Connected.
+    // Thus invoke the authorization service check if not done yet.
+    if (status_ == Status::NotStarted) {
+      callCheck();
```

Comment:
Why do we treat SSL and non-SSL differently? Why not, in both cases, wait until we get the first onData? Then we can pause the pipe and perform auth check. Seems like that would have simpler flow, but I may well be missing something here, so interested to hear your thoughts.

Reply:
That is a very good point/observation. Deferring to `onData()` would definitely simplify the call logic.

Reply:
nice, I didnt think about that